tests/Collection/SearchIndexFunctionalTest.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace MongoDB\Tests\Collection;
+
+use MongoDB\Driver\Cursor;
+use MongoDB\Tests\FunctionalTestCase;
+
+use function bin2hex;
+use function getenv;
+use function random_bytes;
+use function sleep;
+
+/**
+ * Functional tests for the Atlas Search index management.
+ */
+class SearchIndexFunctionalTest extends FunctionalTestCase
+{
+    public function testIndexLifecycle(): void
+    {
+        $atlasUri = getenv('MONGODB_ATLAS_URI');
+
+        if (! $atlasUri) {
+            $this->markTestSkipped('MONGODB_ATLAS_URI environment variable is not set');
+        }
+
+        $this->manager = static::createTestManager($atlasUri);
+
+        $collection = $this->createCollection($this->getDatabaseName(), $this->getCollectionName());
+
+        $name = 'search_index_' . bin2hex(random_bytes(5));
+
+        // Create a search index
+        $createdName = $collection->createSearchIndex(['name' => $name, 'definition' => ['mappings' => ['dynamic' => true]]]);
+        $this->assertSame($name, $createdName);
+
+        // Wait for the index to be ready
+        $count = 0;
+        do {
+            sleep(1);
+            $result = $collection->listSearchIndexes($name, ['typeMap' => ['root' => 'array', 'document' => 'array']]);
+            $this->assertInstanceOf(Cursor::class, $result);
+            $result->rewind();
+            $index = $result->current();
+
+            if ($count++ > 90) {
+                $this->fail('Search index did not become queryable');
+            }
+        } while (! $index['queryable']);
+
+        // Update the search index
+        $collection->updateSearchIndex($name, [
+            'mappings' => ['dynamic' => false],
+        ]);
+
+        // Delete the search index
+        $collection->dropSearchIndex($name);
+
+        $count = 0;
+        do {
+            sleep(1);
+            $result = $collection->listSearchIndexes($name, ['typeMap' => ['root' => 'array', 'document' => 'array']]);
+            $this->assertInstanceOf(Cursor::class, $result);
+            $result->rewind();
+
+            if ($count++ > 90) {
+                $this->fail('Search index was not deleted');
+            }
+        } while ($result->valid());
```

Comment:
I can't use `$result->count()` here because there is an exception when `$result instanceof MongoDB\Model\CachingIterator` and the list of indexes is empty. `MongoDB\Driver\Exception\RuntimeException : Cannot advance a completed or failed cursor.`

Reply:
That seems like a bug in CachingIterator. Would you mind opening a ticket and looking into it? Here's a short script to reproduce it locally (change the URI accordingly):

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$client = new MongoDB\Client('mongodb://localhost:27070/?replicaSet=rs0');
$collection = $client->test->foo;
$collection->drop();

$it = new MongoDB\Model\CachingIterator($collection->find());

var_dump($it);
var_dump($it->count());
```

Reply:
Fix https://github.com/mongodb/mongo-php-library/pull/1118